instance/models/instance.py

```diff
@@ -0,0 +1,436 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app models - Instance
+"""
+
+# Imports #####################################################################
+
+import os
+
+from django.conf import settings
+from django.core.validators import RegexValidator
+from django.db import models
+from django.template import loader
+from django.utils import timezone
+from django_extensions.db.models import TimeStampedModel
+
+from instance import ansible, github
+from instance.gandi import GandiAPI
+from instance.github import fork_name2tuple, get_username_list_from_team
+from instance.log_exception import log_exception
+from instance.repo import clone_configuration_repo
+from instance.models.logging_mixin import LoggerInstanceMixin
+from instance.models.utils import ValidateModelMixin
+
+
+# Constants ###################################################################
+
+PROTOCOL_CHOICES = (
+    ('http', 'HTTP - Unencrypted clear text'),
+    ('https', 'HTTPS - Encrypted'),
+)
+
+gandi = GandiAPI()
+
+
+# Exceptions ##################################################################
+
+class InconsistentInstanceState(Exception):
+    """
+    Indicates that the status of an instance can't be determined
+    """
+    pass
+
+
+# Validators ##################################################################
+
+sha1_validator = RegexValidator(regex='^[0-9a-f]{40}$', message='Full SHA1 hash required')
+
+
+# Models ######################################################################
+
+class Instance(ValidateModelMixin, TimeStampedModel):
+    """
+    Instance - Group of servers running an application made of multiple services
+    """
+    sub_domain = models.CharField(max_length=50)
+    email = models.EmailField(default='contact@example.com')
+    name = models.CharField(max_length=250)
+
+    base_domain = models.CharField(max_length=50, default=settings.INSTANCES_BASE_DOMAIN)
+    protocol = models.CharField(max_length=5, default='http', choices=PROTOCOL_CHOICES)
+
+    last_provisioning_started = models.DateTimeField(blank=True, null=True)
+
+    class Meta:
+        abstract = True
+        unique_together = ('base_domain', 'sub_domain')
+
+    def __str__(self):
+        return '{0.name} ({0.url})'.format(self)
+
+    @property
+    def domain(self):
+        """
+        Instance domain name
+        """
+        return '{0.sub_domain}.{0.base_domain}'.format(self)
+
+    @property
+    def url(self):
+        """
+        Instance URL
+        """
+        return u'{0.protocol}://{0.domain}/'.format(self)
+
+    @property
+    def active_server_set(self):
+        """
+        Returns the subset of `self.server_set` which aren't terminated
+        """
+        return self.server_set.exclude_terminated()
```

Comment:
I found the cloud APIs difficult to abstract. Their semantic varies in ways that have a structural impact on the code using them. For instance some APIs are blocking, others are not, others are optionally and partially blocking. For instance with OpenStack you can create a volume but you can't make that a blocking call which means you then have to pool the volume status until it is available or fails to be. With OpenStack you can however create a server and make that a blocking operation. But it depends which API you're using (the openstack cli does that but not the native nova cli).

I chose to write code that does not abstract the OpenStack behavior at all so that it is abundantly clear to the maintainer that there is no abstraction intended.